### ISSUE TITLE
Reference leak in `test_datetime`

### ISSUE THREAD

**Issue by Eclips4** [python/cpython#124606]
# Bug report

### Bug description:

```python
./python -m test -R 3:3 test_datetime
Using random seed: 2162639794
0:00:00 load avg: 88.54 Run 1 test sequentially in a single process
0:00:00 load avg: 88.54 [1/1] test_datetime
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
XXX 999
test_datetime leaked [9, 9, 9] references, sum=27
test_datetime leaked [3, 3, 3] memory blocks, sum=9
test_datetime failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_datetime

Total duration: 21.9 sec
Total tests: run=1,026 skipped=31
Total test files: run=1/1 failed=1
Result: FAILURE
```

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

macOS

### Linked PRs
* gh-124607

**Comment by vstinner** [comment]
The failing test is `test_fromisoformat_fails_datetime`:

```
$ ./python -m test test_datetime -R 3:3 -m test_fromisoformat_fails_datetime
(...)
test_datetime leaked [9, 9, 9] references, sum=27
(...)
```

**Comment by Eclips4** [comment]
Fixed in #124607.
